### PR TITLE
fix(collection-view): preserve default-order removal survivors

### PR DESCRIPTION
**Pull request by paulfalgout**
Related #328

## What

- Preserve surviving child elements for removal-only updates using Marionette's default collection-order comparator, as well as when ordering is disabled.
- Keep custom comparators, filters, overridden update hooks, deferred children, empty-view transitions, additions, and merges on the full sort/filter/render path.
- Expand unit, packed-package, and Chromium/Firefox/WebKit coverage and update the v5 changelog, CollectionView guide, and migration contract.

## Why

Removing a model cannot change the relative order of surviving children when Marionette orders them by their model's collection index. The existing v5 fast path already preserved survivors when ordering was disabled, but the default `sortWithCollection: true` path still moved all survivors through a document fragment. For a 1,001-row removal this changed 1,000 survivor insertions into work that should be unnecessary and could disturb focus, selection, media, and custom-element connection state.

## Agent-development failure addressed

- Removes a configuration-sensitive performance/correctness trap: an agent can now use the default CollectionView ordering and Marionette's standard collection `update` payload without discovering that every survivor is detached and reinserted on removal.

## Public behavior contract

- Canonical behavior after this PR: an already-rendered, unfiltered CollectionView with no custom comparator destroys removed children without sorting, rerendering, or moving synchronized survivors during a removal-only update.
- Superseded behavior removed, if any: the default collection-order path no longer emits `before:sort`, `sort`, or child-render lifecycles when removal leaves survivor order unchanged.
- Compatibility exception and removal condition, if any: none. Custom comparators, filters, and overridden query/sort/filter hooks deliberately retain the established full path because Marionette cannot prove their side effects or ordering invariants.

## Runtime cost

- [ ] Static or documentation only
- [ ] Development or test only; excluded from production entrypoints
- [x] Changes an existing production path
- [ ] Adds an explicitly opt-in runtime path

Measured impact or explanation:

- Bundle: 75.73 kB cumulative against the 75.80 kB ceiling; no new module or entry point.
- Startup/hot path: the default 1,001-row removal changes from 1,000 survivor `attachHtml` insertions to zero. Local timing reports `collection-view-add-remove` at 39.63 microseconds median and 44.14 microseconds p95.
- Allocations/retention: no new retained objects or allocation-owning runtime path; the update skips an existing fragment/render pass.

## Scope

- Impacts removal-only `CollectionView` collection updates using default collection ordering or disabled ordering.
- Updates generated distributable artifacts and public v5 documentation.
- Does not change reset behavior, custom comparator/filter behavior, add/merge behavior, Application/State contracts, package entry points, or repository settings.

## Deployment Impact

No application or website deployment is performed. The behavior becomes available only in a future Marionette package release; no consumer redeploy is required before then.

## Validation

- [x] Tests cover the acceptance criteria and relevant edge cases.
- [x] Docs, examples, declarations, diagnostics, and rule metadata agree where relevant.
- [x] No private consumer, private fixture, or undocumented context is required.
- [x] I ran the commands listed below and am reporting their actual results.

Commands:

```sh
npm run lint:ci
npm test -- --run --reporter=dot
npm run test:source
npm run test:fixtures
npm run test:browser
npm run test:performance
npm run docs:check
npm run size
npm run performance:timing
npm run build
git diff --check
```

Results: 100 Vitest files / 1,927 tests passed before the final focused addition; the final CollectionView data suite passed 34/34. All packed-package fixtures passed. Both unsorted and default-order 1,001-row removal scenarios passed in Chromium, Firefox, and WebKit. All 106 performance-contract tests passed. Documentation built 48 pages and validated 49 HTML files / 443 internal links. Lint and source validation passed.

## Package and browser impact

- Entry points or install fixtures affected: root runtime behavior only; the existing packed-package removal fixture now validates default ordering.
- Browser contracts affected: survivor node identity, focus, selection, media state, and custom-element connection state are now preserved under default ordering in Chromium, Firefox, and WebKit.
- Production/dev/test module-graph impact: existing `modules/collection-view.js` path only; no new production module or import.

## Rollback or deprecation

- Revert this commit to restore the full sort/filter/render pass for default-order removals. No persisted data or migration rollback is involved.

## Review notes

- The fast-path guards remain intentionally conservative. In particular, method identity checks preserve custom `getComparator`, `getFilter`, `sort`, and `filter` behavior.
- `collection.reset()` is handled separately by `_onCollectionReset()` and still destroys/rebuilds/sorts children.

---
## Summary by cubic
Addresses #328 by fixing removal-only `CollectionView` updates so surviving children stay mounted with default collection ordering, not just when ordering is disabled. Previously the default-order path detached and reinserted every survivor through a document fragment on each removal, which could disturb focus, selection, media, and custom-element connection state.

**Bug Fixes**
- The fast path applies only to removal-only updates with no custom comparator or filter and default `sort`/`filter`/`getComparator`/`getFilter` methods; additions, merges, deferred children, and empty-view transitions keep the full update path.
- Unchanged survivors no longer fire `before:sort`, `sort`, or child-render lifecycles.
- The migration contract now documents that private `_viewComparator` and `_onCollectionUpdate` overrides are not supported extension points.

<sup>Written for commit a1af79a3395cab6986c74c147824342da04133d4. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/marionettejs/marionette/pull/356?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved removal-only CollectionView updates when using default collection ordering or ordering disabled.
  - Preserved surviving child views and their state without unnecessary movement or rerendering.
  - Prevented sort and child-render lifecycle events for unchanged survivors.
  - Retained the standard update path for additions, merges, filters, custom comparators, and overridden behaviors.

- **Documentation**
  - Clarified removal-only update behavior, supported configurations, and migration considerations.

### CODE REVIEW

**Comment by paulfalgout** [comment]
**Approved** at `1b8f5ad9` (the SHA I fetched). CI green, and the bundle **shrinks** 55 B — aggregate 75,728 B against the 75,799 B ceiling, so headroom goes from 16 B to 71 B. Worth noting given how tight that has become: this is a correctness fix that pays for itself.

This closes [#328][i328], one of the two issues opened during the #326 roadmap work.

## The widened predicate is sound

```diff
-    const isUnsorted = this.viewComparator === false ||
-      (!this.viewComparator && !this.sortWithCollection);
+    const isRemovalOrderStable = !this.viewComparator;
```

The justification holds: removal cannot change the *relative* order of survivors under any ordering that is a function of position in the collection. `sortWithCollection: true` orders by collection index, and deleting an element leaves every remaining pair in the same relative order. The old predicate admitted only the ordering-disabled case; the new one admits that plus collection-order, which are exactly the two orderings `!viewComparator` can produce. An overridden `getComparator` is still excluded separately, so a comparator supplied that way cannot slip through.

**The rename is load-bearing, not cosmetic.** `isUnsorted` described the old condition accurately; carrying that name onto the widened check would have made the code assert something false — claiming a `sortWithCollection: true` view is unsorted. `isRemovalOrderStable` names the property that actually licenses the fast path, which is what the next reader needs to evaluate whether some future condition still qualifies.

## The third test is the one that earns its place

`keeps default-order bookkeeping aligned for a later addition`.

A fast path's characteristic failure is not the operation it optimizes — it is the **next** one, when internal state it declined to update turns out to have mattered. Testing only that survivors did not move would pass on an implementation that skips the render and leaves indices or `_children` ordering stale, and the damage would surface on an unrelated later `add` with no obvious connection to the removal that caused it.

Pairing it with `keeps the render path when a custom comparator is active` covers the other direction — proving the fast path declines when it should, not just that it works when it fires. Both directions plus the follow-on operation is the right shape for a conditional optimization.

The remaining guards are intact: `!this.viewFilter` excludes a declared filter, and `children.length === this._children.length` excludes one that is actively hiding children, which is the case a `viewFilter` check alone would miss.

[i328]: https://github.com/marionettejs/marionette/issues/328

File: modules/collection-view.js [hunk 1]
```diff
@@ -207,22 +207,22 @@ assignOwn(CollectionView.prototype, ViewMixin, {
...
-      usesDefaultComparator &&
-      usesDefaultFilterQuery &&
-      usesDefaultSort &&
-      usesDefaultFilter &&
-      isUnsorted &&
+      isDefaultComparator &&
+      isDefaultFilterQuery &&
+      isDefaultSort &&
+      isDefaultFilter &&
+      isRemovalOrderStable &&
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply modules/collection-view.js]: Handled in daa0b4cc with a prototype `_onCollectionUpdate` guard and regression coverage.

File: modules/collection-view.js [hunk 2]
```diff
@@ -207,22 +207,22 @@ assignOwn(CollectionView.prototype, ViewMixin, {
...
-    const usesDefaultSort = this.sort === CollectionView.prototype.sort;
-    const usesDefaultFilter = this.filter === CollectionView.prototype.filter;
-    const isUnsorted = this.viewComparator === false ||
-      (!this.viewComparator && !this.sortWithCollection);
+    const isDefaultComparator = this.getComparator === CollectionView.prototype.getComparator;
+    const isDefaultFilterQuery = this.getFilter === CollectionView.prototype.getFilter;
+    const isDefaultSort = this.sort === CollectionView.prototype.sort;
+    const isDefaultFilter = this.filter === CollectionView.prototype.filter;
+    // Removing models preserves survivor order with collection ordering or no ordering.
+    const isRemovalOrderStable = !this.viewComparator;
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply modules/collection-view.js]: Handled in daa0b4cc by accepting only disabled ordering or the prototype collection-order comparator, with regression coverage.

File: docs/migration-from-v4.md [hunk 3]
```diff
@@ -66,7 +66,7 @@ Status values describe the v5 outcome:
...
+| CollectionView removal-only updates | Every collection update flowed through sort, filter, and child rendering, so removing one model could move every surviving child through a document fragment even when default collection ordering or disabled ordering already preserved their relative order. | A removal-only update on an already-rendered, unfiltered CollectionView with default collection ordering or ordering disabled destroys the removed child without moving or rerendering synchronized visible survivors. Empty-view, deferred-child, custom filter or comparator, add, and merge cases retain the full update path. | Changed | Observe `remove:child` for removals. Do not rely on sort or `render:children` firing when no surviving child needs sorting, rendering, or placement. |
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply docs/migration-from-v4.md]: Handled in daa0b4cc; the migration row now names the default update, sort, filter, and comparator-query requirements.

**Comment by paulfalgout** [comment]
**Approval stands** at `daa0b4cc` — and this push closes a hole I approved past. Correcting that.

## `!this.viewComparator` was not sufficient

I called the previous predicate sound. It was sound for the *declared* comparator; I did not follow `getComparator()` through to what it actually resolves:

```js
getComparator() {
  if (this.viewComparator) { return this.viewComparator; }
  if (!this.sortWithCollection || this.viewComparator === false || !this.collection) { return false; }
  return this._viewComparator;          // ← the instance's, not the prototype's
}
```

So with `viewComparator` falsy and `sortWithCollection: true`, the effective comparator is `this._viewComparator` — an ordinary method a subclass can override. Such a subclass passes `!this.viewComparator` while ordering children by arbitrary logic, and removal is **not** order-stable under arbitrary ordering. The fast path would have skipped a reorder that was genuinely required.

Comparing the resolved value against `CollectionView.prototype._viewComparator` is the right fix: an instance override fails the identity check and the fast path declines. The guard now keys on the ordering actually in effect rather than on the declaration that usually implies it.

`this._onCollectionUpdate === CollectionView.prototype._onCollectionUpdate` closes the same shape one level up, and both have matching tests — `keeps the render path when the collection-order comparator is overridden` and `... when the collection update hook is overridden`. Testing the decline path for each override is what makes these guards verifiable rather than assumed.

The lesson generalizes past this PR: `isDefaultComparator` checks that `getComparator` itself is not overridden, which reads like it covers comparator customization — but it says nothing about what the default implementation *returns*, and that return value passes through another overridable method. Any similar `isDefault*` check on a resolver deserves the same follow-through.

## Bundle: headroom is now 10 bytes

The added guards cost bytes, so this head is **+6 B against master** rather than the −55 B of `1b8f5ad9`. Aggregate 75,789 B against the 75,799 B ceiling — **10 bytes of headroom**, and Bundle size passes.

Correctness is worth the 6 B and I would not trade it back. But at 10 bytes, the next PR that adds any shipped code breaches the ceiling, so the budget decision is effectively due now rather than soon.

**Comment by paulfalgout** [comment]
Addressed the CodeRabbit private-API concern in `d8b5b6c4`. The removal fast path now declines only for documented public `getComparator`, `getFilter`, `sort`, `filter`, `viewComparator`, and `viewFilter` customizations. Tests no longer override `_viewComparator` or `_onCollectionUpdate`, so private internals are not promoted into consumer contracts. Focused and full unit suites, lint, docs, bundle size, and `git diff --check` pass.

**Comment by paulfalgout** [comment]
**Approval stands** at `d8b5b6c4`, and I want to walk back my framing from two comments ago rather than leave it standing.

I called the `getComparator()`-resolution guard "closing a real hole." The **mechanism** I described is real and still is: `getComparator()` returns `this._viewComparator`, that is an ordinary prototype method, and a subclass overriding it with arbitrary ordering passes `!this.viewComparator` and takes the fast path. Nothing on this head changes that.

What I did not weigh is whether that path is *supported*. `_viewComparator` and `_onCollectionUpdate` are underscore-prefixed, and the convention in this very predicate is consistent about which methods get identity guards:

```
isDefaultComparator   → getComparator     (public)
isDefaultFilterQuery  → getFilter         (public)
isDefaultSort         → sort              (public)
isDefaultFilter       → filter            (public)
```

Four guards, all on public extension points. The version I praised added two more on private methods, which made it the outlier rather than the correction. Guarding every private method a public one delegates to is unbounded — the same argument would demand identity checks across most of the prototype.

So this revert restores the established boundary, and the migration row was narrowed to match (`overridden query methods` rather than `overridden update or query methods`), which is consistent rather than sloppy. I should have asked "is this a supported override point?" before calling it a hole.

## One thing worth recording

The previous head carried two tests — `keeps the render path when the collection-order comparator is overridden` and `... when the collection update hook is overridden` — that asserted the opposite position, and they are deleted here with nothing stating why. The reasoning is now implicit in a naming convention.

If the position is "overriding `_viewComparator` or `_onCollectionUpdate` is unsupported, so the fast path need not defend against it," that is defensible and worth one line — in the migration row, or a comment on the predicate. Otherwise the next person to notice the mechanism reruns this exact loop, and the only artifact of it having been considered is a deleted test.

## Bundle

**−105 B** against master, better than both earlier versions (−55 B unguarded, +6 B guarded). Aggregate 75,678 B, **headroom 121 B**. With #357's −617 B also pending, the ceiling pressure I flagged earlier is easing from both directions.

**Comment by paulfalgout** [comment]
**Approval stands** at `353e13e3`. Documentation only; the −105 B figure is unchanged.

This records the position I said was implicit:

> Private `_viewComparator` and `_onCollectionUpdate` overrides are not supported extension points; use the documented public options and methods.

That is the right resolution and the right place for it. The reasoning was previously recoverable only from a naming convention, with the two deleted tests as the sole evidence anyone had considered it. Now the boundary is stated where a consumer looks when their subclass stops behaving as expected, and the deleted tests are explained rather than merely absent.

It also generalizes past this fast path: it tells a reader that underscore-prefixed methods on `CollectionView` are not extension surface at all, which is worth more than a note scoped to these two.

File: test/unit/collection-view/collection-view-data.spec.js [hunk 4]
```diff
@@ -387,7 +387,47 @@ describe('CollectionView Data', function() {
       expect([...myCollectionView.el.children]).to.deep.equal(survivorNodes);
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply test/unit/collection-view/collection-view-data.spec.js:387]: This behavior is intentional. `_viewComparator` and `_onCollectionUpdate` are private internals, not supported extension points, so the public-contract fast path does not guard private overrides. Exact head `353e13e3` now records that boundary explicitly in the v4 migration row and directs consumers to documented public options and methods. The focused and full unit suites, lint, docs validation, and bundle-size validation passed before this documentation follow-up; `npm run docs:check` and `git diff --check` pass on the exact head.

**Comment by paulfalgout** [comment]
**Approval stands** at `402aac12`. `modules/collection-view.js` is byte-identical to `353e13e3` and the unsupported-override note is intact — this is a rebase, nothing to re-review.

One correction to the numbers in my previous comment, since master moved underneath them. #357 has now merged as `dacc74f7`, which removed the feature registry and took master from 75,783 B to **75,166 B (headroom 633 B)**. This branch is still based on `f4eac6c2`, the pre-#357 master, and still carries `runtime/features.js`.

So the honest figures:

- **This PR's own contribution: −105 B**, measured against its actual base rather than current master. Unchanged from before.
- Comparing it directly against `dacc74f7` yields a meaningless +512 B, because that difference is dominated by #357's removal being present in master and absent here.

Once rebased, the −105 B applies to 75,166 B, landing near **75,061 B with roughly 738 B of headroom**. No budget concern either way — worth stating only because a naive master-versus-branch comparison right now looks like a 512 B regression and is not one.

**Comment by paulfalgout** [comment]
**Approval stands** at `a1af79a3`, now rebased onto current master `dacc74f7`. `modules/collection-view.js` is byte-identical to `402aac12`, so this is purely a base change.

That makes the measurement I flagged as blocked last comment finally clean:

```
master dacc74f7   75,166 B
#356  a1af79a3    75,128 B    −38 B    headroom 671 B
```

The −38 B differs from the −105 B I quoted earlier, and the earlier figure was not wrong — it was measured against `f4eac6c2`, where the surrounding code still contained the feature registry. Removing that registry in #357 changed what compresses alongside this change, so the same source edit now nets differently. Worth noting because it is a reminder that Brotli deltas are properties of a *pair*, not of a patch: quoting one without its base invites exactly the confusion the +512 B figure caused.

Either way it is net-negative and there is no budget question. `Bundle size` was still running when I looked.
